### ISSUE TITLE
UnicodeDecode Error 

### ISSUE THREAD

**Issue by BuonOmo** [codeclimate/python-test-reporter#37]
Hi,

I’m encountering a Python error every time I launch command:

    codeclimate-test-reporter --debug --token [my_token]

You can reproduce the error this way:

    git clone https://github.com/buonomo/inf8007
    cd inf8007
    pip install -r requirements.txt
    python -m nltk.downloader "stopwords"
    coverage run test_td2.py
    codeclimate-test-reporter --debug --token [my_token, ask if needed]

Below, the [stdout](#stdout) and [stderr](#stderr) I obtained.

# <a id=stdout></a>STDOUT
> codeclimate-test-repoter 0.2.0
Requests 2.11.1
Python 3.5.2 (default, Nov 17 2016, 17:05:23) 
[GCC 5.4.0 20160609]
/usr/bin/python3
Linux 4.4.0-62-genericProcessing file path=td2.py
Processing file path=test_td2.py
Processing file path=/usr/lib/python3/dist-packages/six.py
Processing file path=/usr/local/lib/python3.5/dist-packages/nltk/__init__.py
Processing file path=/usr/local/lib/python3.5/dist-packages/nltk/collections.py
Processing file path=/usr/local/lib/python3.5/dist-packages/nltk/collocations.py
Processing file path=/usr/local/lib/python3.5/dist-packages/nltk/compat.py
Processing file path=/usr/local/lib/python3.5/dist-packages/nltk/data.py
Processing file path=/usr/local/lib/python3.5/dist-packages/nltk/decorators.py
Processing file path=/usr/local/lib/python3.5/dist-packages/nltk/downloader.py
Processing file path=/usr/local/lib/python3.5/dist-packages/nltk/featstruct.py
Processing file path=/usr/local/lib/python3.5/dist-packages/nltk/grammar.py
Processing file path=/usr/local/lib/python3.5/dist-packages/nltk/help.py
Processing file path=/usr/local/lib/python3.5/dist-packages/nltk/internals.py
Processing file path=/usr/local/lib/python3.5/dist-packages/nltk/jsontags.py
Processing file path=/usr/local/lib/python3.5/dist-packages/nltk/lazyimport.py
Processing file path=/usr/local/lib/python3.5/dist-packages/nltk/probability.py
Processing file path=/usr/local/lib/python3.5/dist-packages/nltk/text.py
Processing file path=/usr/local/lib/python3.5/dist-packages/nltk/toolbox.py
Processing file path=/usr/local/lib/python3.5/dist-packages/nltk/tree.py
Processing file path=/usr/local/lib/python3.5/dist-packages/nltk/treetransforms.py
Processing file path=/usr/local/lib/python3.5/dist-packages/nltk/util.py
Processing file path=/usr/local/lib/python3.5/dist-packages/nltk/wsd.py
Processing file path=/usr/local/lib/python3.5/dist-packages/nltk/ccg/__init__.py
Processing file path=/usr/local/lib/python3.5/dist-packages/nltk/ccg/api.py
Processing file path=/usr/local/lib/python3.5/dist-packages/nltk/ccg/chart.py
Processing file path=/usr/local/lib/python3.5/dist-packages/nltk/ccg/combinator.py
Processing file path=/usr/local/lib/python3.5/dist-packages/nltk/ccg/lexicon.py
Processing file path=/usr/local/lib/python3.5/dist-packages/nltk/ccg/logic.py
Processing file path=/usr/local/lib/python3.5/dist-packages/nltk/chunk/__init__.py
Processing file path=/usr/local/lib/python3.5/dist-packages/nltk/chunk/api.py
Processing file path=/usr/local/lib/python3.5/dist-packages/nltk/chunk/regexp.py
Processing file path=/usr/local/lib/python3.5/dist-packages/nltk/chunk/util.py
Processing file path=/usr/local/lib/python3.5/dist-packages/nltk/classify/__init__.py
Processing file path=/usr/local/lib/python3.5/dist-packages/nltk/classify/api.py
Processing file path=/usr/local/lib/python3.5/dist-packages/nltk/classify/decisiontree.py
Processing file path=/usr/local/lib/python3.5/dist-packages/nltk/classify/maxent.py
Processing file path=/usr/local/lib/python3.5/dist-packages/nltk/classify/megam.py
Processing file path=/usr/local/lib/python3.5/dist-packages/nltk/classify/naivebayes.py
Processing file path=/usr/local/lib/python3.5/dist-packages/nltk/classify/positivenaivebayes.py
Processing file path=/usr/local/lib/python3.5/dist-packages/nltk/classify/rte_classify.py
Processing file path=/usr/local/lib/python3.5/dist-packages/nltk/classify/scikitlearn.py
Processing file path=/usr/local/lib/python3.5/dist-packages/nltk/classify/senna.py
Processing file path=/usr/local/lib/python3.5/dist-packages/nltk/classify/tadm.py
Processing file path=/usr/local/lib/python3.5/dist-packages/nltk/classify/textcat.py
Processing file path=/usr/local/lib/python3.5/dist-packages/nltk/classify/util.py
Processing file path=/usr/local/lib/python3.5/dist-packages/nltk/classify/weka.py
Processing file path=/usr/local/lib/python3.5/dist-packages/nltk/cluster/__init__.py
Processing file path=/usr/local/lib/python3.5/dist-packages/nltk/cluster/api.py
Processing file path=/usr/local/lib/python3.5/dist-packages/nltk/cluster/em.py
Processing file path=/usr/local/lib/python3.5/dist-packages/nltk/cluster/gaac.py
Processing file path=/usr/local/lib/python3.5/dist-packages/nltk/cluster/kmeans.py
Processing file path=/usr/local/lib/python3.5/dist-packages/nltk/cluster/util.py
Processing file path=/usr/local/lib/python3.5/dist-packages/nltk/corpus/__init__.py
Processing file path=/usr/local/lib/python3.5/dist-packages/nltk/corpus/util.py
Processing file path=/usr/local/lib/python3.5/dist-packages/nltk/corpus/reader/__init__.py
Processing file path=/usr/local/lib/python3.5/dist-packages/nltk/corpus/reader/aligned.py
Processing file path=/usr/local/lib/python3.5/dist-packages/nltk/corpus/reader/api.py
Processing file path=/usr/local/lib/python3.5/dist-packages/nltk/corpus/reader/bnc.py
Processing file path=/usr/local/lib/python3.5/dist-packages/nltk/corpus/reader/bracket_parse.py
Processing file path=/usr/local/lib/python3.5/dist-packages/nltk/corpus/reader/categorized_sents.py
Processing file path=/usr/local/lib/python3.5/dist-packages/nltk/corpus/reader/chasen.py
Processing file path=/usr/local/lib/python3.5/dist-packages/nltk/corpus/reader/childes.py
Processing file path=/usr/local/lib/python3.5/dist-packages/nltk/corpus/reader/chunked.py
Processing file path=/usr/local/lib/python3.5/dist-packages/nltk/corpus/reader/cmudict.py
Processing file path=/usr/local/lib/python3.5/dist-packages/nltk/corpus/reader/comparative_sents.py
Processing file path=/usr/local/lib/python3.5/dist-packages/nltk/corpus/reader/conll.py
Processing file path=/usr/local/lib/python3.5/dist-packages/nltk/corpus/reader/crubadan.py
Processing file path=/usr/local/lib/python3.5/dist-packages/nltk/corpus/reader/dependency.py
Processing file path=/usr/local/lib/python3.5/dist-packages/nltk/corpus/reader/framenet.py
Processing file path=/usr/local/lib/python3.5/dist-packages/nltk/corpus/reader/ieer.py
Processing file path=/usr/local/lib/python3.5/dist-packages/nltk/corpus/reader/indian.py
Processing file path=/usr/local/lib/python3.5/dist-packages/nltk/corpus/reader/ipipan.py
Processing file path=/usr/local/lib/python3.5/dist-packages/nltk/corpus/reader/knbc.py
Processing file path=/usr/local/lib/python3.5/dist-packages/nltk/corpus/reader/lin.py
Processing file path=/usr/local/lib/python3.5/dist-packages/nltk/corpus/reader/mte.py
Processing file path=/usr/local/lib/python3.5/dist-packages/nltk/corpus/reader/nkjp.py
Processing file path=/usr/local/lib/python3.5/dist-packages/nltk/corpus/reader/nombank.py
Processing file path=/usr/local/lib/python3.5/dist-packages/nltk/corpus/reader/nps_chat.py
Processing file path=/usr/local/lib/python3.5/dist-packages/nltk/corpus/reader/opinion_lexicon.py
Processing file path=/usr/local/lib/python3.5/dist-packages/nltk/corpus/reader/panlex_lite.py
Processing file path=/usr/local/lib/python3.5/dist-packages/nltk/corpus/reader/pl196x.py
Processing file path=/usr/local/lib/python3.5/dist-packages/nltk/corpus/reader/plaintext.py
Processing file path=/usr/local/lib/python3.5/dist-packages/nltk/corpus/reader/ppattach.py
Processing file path=/usr/local/lib/python3.5/dist-packages/nltk/corpus/reader/propbank.py
Processing file path=/usr/local/lib/python3.5/dist-packages/nltk/corpus/reader/pros_cons.py
Processing file path=/usr/local/lib/python3.5/dist-packages/nltk/corpus/reader/reviews.py
Processing file path=/usr/local/lib/python3.5/dist-packages/nltk/corpus/reader/rte.py
Processing file path=/usr/local/lib/python3.5/dist-packages/nltk/corpus/reader/semcor.py
Processing file path=/usr/local/lib/python3.5/dist-packages/nltk/corpus/reader/senseval.py
Processing file path=/usr/local/lib/python3.5/dist-packages/nltk/corpus/reader/sentiwordnet.py
Processing file path=/usr/local/lib/python3.5/dist-packages/nltk/corpus/reader/sinica_treebank.py
Processing file path=/usr/local/lib/python3.5/dist-packages/nltk/corpus/reader/string_category.py
Processing file path=/usr/local/lib/python3.5/dist-packages/nltk/corpus/reader/switchboard.py
Processing file path=/usr/local/lib/python3.5/dist-packages/nltk/corpus/reader/tagged.py
Processing file path=/usr/local/lib/python3.5/dist-packages/nltk/corpus/reader/timit.py
Processing file path=/usr/local/lib/python3.5/dist-packages/nltk/corpus/reader/toolbox.py
Processing file path=/usr/local/lib/python3.5/dist-packages/nltk/corpus/reader/twitter.py
Processing file path=/usr/local/lib/python3.5/dist-packages/nltk/corpus/reader/udhr.py
Processing file path=/usr/local/lib/python3.5/dist-packages/nltk/corpus/reader/util.py
Processing file path=/usr/local/lib/python3.5/dist-packages/nltk/corpus/reader/verbnet.py
Processing file path=/usr/local/lib/python3.5/dist-packages/nltk/corpus/reader/wordlist.py
Processing file path=/usr/local/lib/python3.5/dist-packages/nltk/corpus/reader/wordnet.py
Processing file path=/usr/local/lib/python3.5/dist-packages/nltk/corpus/reader/xmldocs.py
Processing file path=/usr/local/lib/python3.5/dist-packages/nltk/corpus/reader/ycoe.py

# <a id=stderr></a>STDERR

> Traceback (most recent call last):
  File "/usr/local/bin/codeclimate-test-reporter", line 9, in <module>
    load_entry_point('codeclimate-test-reporter==0.2.0', 'console_scripts', 'codeclimate-test-reporter')()
  File "/usr/local/lib/python3.5/dist-packages/codeclimate_test_reporter/__main__.py", line 11, in run
    sys.exit(runner.run())
  File "/usr/local/lib/python3.5/dist-packages/codeclimate_test_reporter/components/runner.py", line 49, in run
    exit_status = reporter.run()
  File "/usr/local/lib/python3.5/dist-packages/codeclimate_test_reporter/components/reporter.py", line 36, in run
    payload = formatter.payload()
  File "/usr/local/lib/python3.5/dist-packages/codeclimate_test_reporter/components/formatter.py", line 20, in payload
    source_files = self.__source_files()
  File "/usr/local/lib/python3.5/dist-packages/codeclimate_test_reporter/components/formatter.py", line 58, in __source_files
    payload = file_coverage.payload()
  File "/usr/local/lib/python3.5/dist-packages/codeclimate_test_reporter/components/file_coverage.py", line 18, in payload
    "blob_id": self.__blob(),
  File "/usr/local/lib/python3.5/dist-packages/codeclimate_test_reporter/components/file_coverage.py", line 44, in __blob
    contents = self.__file_body()
  File "/usr/local/lib/python3.5/dist-packages/codeclimate_test_reporter/components/file_coverage.py", line 51, in __file_body
    self.file_body = open(self.__filename(), "r", encoding="utf-8-sig").read()
  File "/usr/lib/python3.5/codecs.py", line 321, in decode
    (result, consumed) = self._buffer_decode(data, self.errors, final)
  File "/usr/lib/python3.5/encodings/utf_8_sig.py", line 69, in _buffer_decode
    return codecs.utf_8_decode(input, errors, final)
UnicodeDecodeError: 'utf-8' codec can't decode byte 0xc6 in position 6000: invalid continuation byte

**Comment by maxjacobson** [comment]
@BuonOmo Hi, thanks for opening this issue. I appreciate the level of detail and reproduction case. I tried to reproduce, but got this issue:

```
$ coverage test_td2.py    
Unknown command: 'test_td2.py'
Use 'coverage help' for help.
```

Tried using `coverage run test_td2.py` and got this issue:

```
$ coverage run test_td2.py
Traceback (most recent call last):
  File "test_td2.py", line 29, in <module>
    class TestParser(TestCase):
  File "test_td2.py", line 30, in TestParser
    parser = Parser('french')
  File "/Users/maxjacobson/src/oss-interesting/inf8007/td2.py", line 54, in __init__
    self.stopwords = set(stopwords.words(language))
  File "/usr/local/lib/python2.7/site-packages/nltk/corpus/util.py", line 116, in __getattr__
    self.__load()
  File "/usr/local/lib/python2.7/site-packages/nltk/corpus/util.py", line 81, in __load
    except LookupError: raise e
LookupError: 
**********************************************************************
  Resource u'corpora/stopwords' not found.  Please use the NLTK
  Downloader to obtain the resource:  >>> nltk.download()
  Searched in:
    - '/Users/maxjacobson/nltk_data'
    - '/usr/share/nltk_data'
    - '/usr/local/share/nltk_data'
    - '/usr/lib/nltk_data'
    - '/usr/local/lib/nltk_data'
**********************************************************************
```

Happy to dig into why you're seeing this issue once I'm able to reproduce.

**Comment by BuonOmo** [comment]
Ok sorry, i forgot about nltk dependencies, let me update my question. I’ve made a fork to see if I could spot the error, but I can’t yet!

**Comment by BuonOmo** [comment]
I found a way to make it work, I ignore external files with my coverage :

    coverage run --include='./*' test_td2.py

However this doesn’t mean the error is not there!

**Comment by maxjacobson** [comment]
OK, after doing a little digging, it looks like that final line of debug output provides a clue:

```
 Processing file path=/usr/local/lib/python3.5/dist-packages/nltk/corpus/reader/ycoe.py
```

If you inspect that source file, it seems like it isn't validly encoded as a UTF-8 string, so when we try to read it as a UTF-8 encoded string, we get an error.

I suppose one option would be to skip files with invalid encoding. What do you think about that?

**Comment by BuonOmo** [comment]
@maxjacobson I think its great idea, and you should mention those files in user’s shell prompt. Shall I make a PR about it ?

**Comment by maxjacobson** [comment]
That would be great!

**Comment by BuonOmo** [comment]
I did not have time to test it yet but i will submit my PR tonight!

**Comment by BuonOmo** [comment]
After deeper reflexion, rather than excluding file that are not utf-8, wouldn’t it be better to check their encoding and parsing them ?

Either using chardet or cchardet, or simply by parsing python `# -*- coding` prefix.

**Comment by maxjacobson** [comment]
Hmm... Sure, that sounds good too if it can be done reliably

**Comment by BuonOmo** [comment]
Maybe not reliably for all encoding but I think we will cover the huge majority of cases if we try (in this order):
- utf8
- check python encoding
- latin 1
- throw error if still not handled

What do you think ?

**Comment by maxjacobson** [comment]
In the example you cited, the file began with this comment:

```
# -*- coding: iso-8859-1 -*-
```

The idea would be to check for the presence of such a comment, extract the encoding, and then use that encoding to read the file? We would need to read the file in order to detect the comment, no?

**Comment by BuonOmo** [comment]
Of course, but we can use only readline to read lines one by one (and stop at first none blank or comment line). Then we would avoid encoding issues

**Comment by BuonOmo** [comment]
here’s a snippet

```python
with open('my_file') as stream:
    for line in stream:
        if not (line.strip() == '' or line.startswith('#')):
             break
        regex_match = re.match(r"# -\*- coding: (.+) -\*-",line).groups()
        if len(regex_match) > 0:
            encoding = regex_match[0]
```

**Comment by maxjacobson** [comment]
OK it sounds like an improvement to me :)

**Comment by dblandin** [comment]
This was addressed in #38. 👌
